NeoML/include/NeoML/Dnn/DnnHead.h

```diff
@@ -0,0 +1,64 @@
+/* Copyright © 2024 ABBYY
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+--------------------------------------------------------------------------------------------------------------*/
+
+#pragma once
+
+#include <memory>
+#include <initializer_list>
+#include <NeoML/NeoML.h>
+#include <NeoML/Dnn/Dnn.h>
+#include <NeoML/Dnn/Layers/CompositeLayer.h>
+
+namespace NeoML {
+
+class CDnnHeadAdapterLayer;
+
+template <typename T>
+class CLayerWrapper;
+
+struct CHeadStruct : public IObject {
+	CHeadStruct(CDnn* net) : dnn(net) {}
+	CDnn* GetInnerDnn() { return dnn; };
+private:
+	~CHeadStruct() { if (dnn != nullptr) { delete dnn; dnn = nullptr; } }
+	CDnn* dnn;
+	CObjectArray<CDnnHeadAdapterLayer> connections;
+
+	friend class CDnnHeadAdapterLayer;
+};
+
+template <typename... Ts>
+CHeadStruct* CDnnHead(CRandom random, IMathEngine& mathEngine, CLayerWrapper<Ts>... linearWrappers)
```

Comment:
This function should be the unique public constructor for the `CDnnHead` class.
Any other methods should be private, only friends should be able change something inside.
Make friend `optimizeDnnAdapter` function